extensions/ql-vscode/src/telemetry.ts

```diff
@@ -0,0 +1,159 @@
+  if (!reporter) {
+    return;
+  }
+  const status = !error
+    ? CommandCompletion.Success
+    : error instanceof UserCancellationException
+      ? CommandCompletion.Cancelled
+      : CommandCompletion.Failed;
+
+  const isCanary = (!!CANARY_FEATURES.getValue<boolean>()).toString();
+
+  reporter.sendTelemetryEvent(
+    'command-usage',
+    {
+      name,
+      status,
+      isCanary
+    },
+    { executionTime }
+  );
+
+  // if this is a true error, also report it
+  if (status === CommandCompletion.Failed && error) {
+    const redactedError = {
+      name: error.name,
+      stack: error.stack,
+      message: '<MESSAGE REDACTED>'
+    };
+    reporter.sendTelemetryException(
+      redactedError,
+      {
+        type: 'command-usage',
+        name,
+        status,
+        isCanary
+      },
+      { executionTime }
+    );
+  }
+}
+
+function registerListener(extension: Extension<any>, ctx: ExtensionContext) {
+  if (!listener) {
+    listener = workspace.onDidChangeConfiguration(async e => {
+      if (e.affectsConfiguration('codeQL.telemetry.enableTelemetry')) {
+        await initializeTelemetry(extension, ctx);
+      }
+
+      // TODO: re-request telemetry so that users can see the dialog again.
+      // Re-request if codeQL.canary is being set to `true` and telemetry
+      // is not currently enabled.
+      if (
+        e.affectsConfiguration('codeQL.canary') &&
+        CANARY_FEATURES.getValue() &&
+        !ENABLE_TELEMETRY.getValue()
+      ) {
+        await ctx.globalState.update('telemetry-request-viewed', false);
+        await requestTelemetryPermission(ctx);
+      }
+    });
+    ctx.subscriptions.push(listener);
+  }
+}
+
+async function requestTelemetryPermission(ctx: ExtensionContext) {
+  if (!ctx.globalState.get('telemetry-request-viewed')) {
+    // if global telemetry is disabled, avoid showing the dialog or making any changes
+    let result = undefined;
+    if (workspace.getConfiguration().get<boolean>('telemetry.enableTelemetry')) {
+      // Extension won't start until this completes. So, set a timeout here in order
+      // to ensure the extension continues even if the user doesn't make a choice.
+      result = await Promise.race([
+        showBinaryChoiceDialog(
+          'Do we have your permission to collect anonymous usage statistics help us improve CodeQL for VSCode? See [TELEMETRY.md](https://github.com/github/vscode-codeql/blob/93831e28597bb08567db9844737efd9ff188fc3a/.github/TELEMETRY.md)',
+          { modal: false }
+        ),
+        new Promise(resolve => setTimeout(resolve, 10000))
+      ]) as boolean | undefined;
+    }
+    if (result !== undefined) {
+      await Promise.all([
+        ctx.globalState.update('telemetry-request-viewed', true),
+        ENABLE_TELEMETRY.updateValue<boolean>(result, ConfigurationTarget.Global),
+      ]);
+    }
+  }
+}
+
+// Exported for testing
+export function _dispose() {
```

Comment:
It is a little strange to me to only call this during testing.
I see that `listener` and `reporter` are tracked in `ctx.subscriptions` and disposed when the extension is deactivated, but perhaps we can make that more explicit.

What do you think about organising this file into a `Telemetry` class that extends `DisposableObject`, that contains `listener` and `reporter` fields that are created explicitly at construction, can be reset if necessary, and are tracked for disposal? Bit like what we do for the server clients. Then we only push the `Telemetry` object to `ctx.subscriptions`.

Reply:
I'll try this out.